Readme.md
## Gestor de clases.  
  
```  
Copyright (c) 2019 Damián Cabello. All rights reserved.  
  
This work is licensed under the terms of the MIT license. For a copy, see [here](https://opensource.org/licenses/MIT).  
```  
Gestor de aulas web creado en PHP con llamadas a una BD, necesita tener  
un usuario en MySQL con permisos para la ejecucción de comandos.  
  
> El usuario y contraseña de deben ir en el fichero ../core/dao.php  
> Por defecto el usuario es www-data y la contraseña usuario.  
  
  



   
       
  
## TODO  
  
 - [x] Hacer el registro funcional.  
 - [x] Hacer jerarquía de privilegios al sistema. Ya implementado en la base de datos.
   
|ADMIN|USER  |  
|--|--|  
| true (1) |false (0)|  
  
 - [x] ¿Posible refactorización del navbar?  
 - [x] Migrar el proyecto a mi dominio para ver un ejemplo online.  
 - [ ] Posibilidad de buscar aulas.  
 - [x] Posibilidad de reservar aulas (en sus seis tramos)  
 - [ ] Editar aulas ya creadas.  
 - [x] Deshabilitar aulas del sistema.  
 - [x] Consultar reservas de aulas.  
 - [x] Crear un form de creación de aulas.  
 - [ ] Añadir el nombre del usuario al navbar  
 - [ ] Añadir pestaña settings.  
 - [ ] Comentar el codigo.  
 - [ ] Hacer un A cerca de "normal".  
 - [ ] Cambiar la imagen por un patrón.  
 - [ ] AL ACABAR: ¿Refactorizar?.
 - [x] Añadir firma.
 - [ ] El footer no es sticky.
 - [x] -Hacer el changelog como dios manda.

changelog.md
# Changelog  
All notable changes to this project will be documented in this file.  
  
This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) to format this changelog.

## [0.9.0] - 2019-02-15

### Added
- Añadida la vista con distintos permisos en listar aulas.
- Añadido el campo para anular reserva en listar reservas.
- Creado el fichero `view/openformListaulas.php`.
- Creada la tabla log.


## [0.8.1] - 2019-02-13

### Added
- Creado el fichero `versionLog.md`.

### Removed
- Contenido de `readme.md` .

## [0.8.0] - 2019-02-13
### Changed
- Añadido un nuevo campo `gestionaulas.nreserva` en la base de datos.

### Added
- Botón de borrar reserva.
- Gestión de admin en listar reservas.
- Creado el fichero `versionLog.md`.

### Removed
- Contenido de `view/acercaDe.php` .

### Fixed
- Mensaje al administrador al borrar la reserva no mostraba el nombre del usuario, sino del aula, esto se ha arreglado.


## [0.7.0] - 2019-02-11 
### Added  
- Creado el archivo `pages/listReservas.php`. 
- Añadido el navbar V2.
- Añadido en la base de datos el campo aula.deshablitada.
 - Agregada actualización de la base de datos.
- Los botones "habilitar aula" y "deshabilitar aula"  son completamente funcionales.
### Changed  
- Refactorizados los estilos para hacer el footer dinámico.   
- Añadida distinta vista segun el rango del usuario, si es un **administrador** muestra **TODAS las reservas**, si es un **usuario** normal, muestra **SOLO SUS reservas**.
- Las aulas deshabilitadas se muestran en la vista de reservar en rojo y deshabilitadas, y en el listado de aulas se muestra el boton "habilitar".

### Fixed   
- Recogido el tipo de usuario desde la base de datos (administrador o usuario) mediante una llamada a `usuario.admin`.  
- El botón borrar funciona correctamente.
### Deprecated
- El fichero `readme.md` solo va a contener información del proyecto, se creará un fichero de versiones.
## [0.6.0] - 2019-05-02 
### Added  
- Traducida la cabecera de listar aulas. 
- Implementado el campo admin de tipo TINYINT(1) en la tabla usuario.  
- Creado el form para añadir aulas. 
- Añadido el archivo `pages/reservaAula.php`.   
- Añadido el archivo `view/reservaForm.php`. 
- Creada la tabla reserva en la base de datos.  
- Posibilidad de reservar un aula.  
### Changed  
- Refactorizados algunos estilos.  
- Creados los estilos en  `.baseStyleForm select` y `.baseStyleForm select:focus` en `baseStyle.css`.
### Removed  
- Refactorizada la base de datos, ya no existen las tablas `login` y `user` ahora solo existe la tabla `user` con más campos.
- Eliminado `view/head_noboostrap.php` con  las ultimas refactorizaciones ya no era necesario. 
- Eliminado `anadirAulaBD.php` ahora se hace por POST hacia el mismo php.  
### Fixed   
- Se pueden crear usuarios en la base de datos ya funciona el formulario de registro. 
- En el listado de todas las aulas se muestran solo 50 caracteres en la descripción. 
- Añadido el campo todo el día para evitar que se tenga que reservar el mismo aula 6 veces en la misma fecha.  

## [0.2.0] - 2019-01-29 
### Added  
- Se añade la posibilidad de cambiar de idioma (por ahora español e inglés).  
- Se crea métodos para añadir aulas al sistema. 
- Se crea una vista (aún sin estilos) para los métodos.  
- Se añade en la base de datos el campo `admin` a la tabla `usuario` para futuras mejoras.
- Se crea la vista del formulario de registro en `registerForm.php`.  
### Changed  
- Se vuelve a refactorizar para que los strings cambien todos según el idioma.


## [0.1.1] - 2019-01-27 
### Added  
- Se crean las carpetas assets, core, pages y view. 
- Se crea la vista del navbar en el fichero `view/mainnav.php`.
- Se crea la vista del registro en el fichero `view/register.php`.
### Changed  
- Refactorización : se refactoriza todo para separar la vista del controlador.  
         
- Se mueven `dao.php` y `app.php`a la carpeta `core`.  
 - Las carpetas`img`, `css`y `js` se mueven a la capeta `assets`.  
 
## [0.1.0] - 2019-01-26 
### Added  
   - Creación de métodos para **listar todas las aulas**.  
   - Vista para mostrar los datos recibidos.  
   - Botón cerrar sesión con su correspondiente php.  

## [0.0.2] - 2019-01-26 
### Added  
- Creación del fichero `Lobby.php`.
- Creación del main navbar dentro del lobby. 
### Fixed  
- Ahora ya se puede loguear al sistema mediante un formulario almacenado en la base de datos.

## [0.0.1] - 2019-01-25 


### Added  
- Creación del primer modelo de la base de datos, posiblemente cambie.  
- Creación del fichero `dao.php`.
- Creación del fichero `app.php`.
- Creación del fichero `login.php`.



core/dao.php
<?php
include_once("../.idea/psw.php");
define('DATABASE', 'gestionaulas');
define('HOST', 'localhost');
define('USER', 'root');
define('PASS', '');
//define('PASS', $psw);

define('DSN', "mysql:host=" . HOST . ";dbname=" . DATABASE);

// tablas

define('TABLE_USER', 'usuario');
define('COLUMN_LOGIN_USUARIO', 'usuario');
define('COLUMN_LOGIN_CONTRASENIA', 'password');


define('TABLE_AULA', 'aula');

define('TABLE_RESERVE', 'reserva');



class Dao
{
    private $conn;
    public $error;

    function __construct()
    {
        try {
            $this->conn = new PDO(DSN, USER, PASS);
        } catch (PDOException $e) {
            $this->error = "Database conection error: " . $e->getMessage();
        }
    }

    function isConected()
    {
        return isset($this->conn);
    }

    function authenticate($user, $password)
    {
        if (!isset($user, $password)) {
            return false;
        }
        $sql = "SELECT usuario, password FROM ".TABLE_USER." WHERE usuario = '".$user."'  AND password = SHA2(\"".$password."\",512)";
        $stmt = $this->conn->query($sql);
        if ($stmt->rowCount() == 1) {
            return true;
        }
        return false;
    }

    /**
     * Métodos de gestión de aulas
     */

    function insertAula($nombre, $descripcion,$posicion,$estic,$pccount)
    {
        try {
            $sql = "INSERT INTO " . TABLE_AULA . " (nombre, descripcion, ubicacion, tic, numordenadores) VALUES ('" . $nombre . "','" . $descripcion . "','" . $posicion . "','" . $estic . "','" . $pccount . "')";
            if($this->insertaLog("INSERT",$_SESSION['user']." creó un aula.", $_SESSION['user'], $sql)) {
                $resultset = $this->conn->query($sql);
                return $resultset;
            }
        } catch (PDOException $e) {
            $this->error = $e->getMessage();
        }
    }

    function deletetreserva($numreserva)
    {
        try {
            $sql = "DELETE FROM ".TABLE_RESERVE." WHERE nreserva ='".$numreserva."'";
            if($this->insertaLog("DELETE",$_SESSION['user']." canceló una reserva.", $_SESSION['user'], $sql)) {
                $resultset = $this->conn->query($sql);

                return $resultset;
            }
        } catch (PDOException $e) {
            $this->error = $e->getMessage();
        }
    }

    function deletetAula($nombre)
    {
        try {
            $sql = "DELETE FROM ".TABLE_AULA." WHERE nombre ='".$nombre."'";
            if($this->insertaLog("DELETE",$_SESSION['user']." eliminó un aula.", $_SESSION['user'], $sql)) {
                $resultset = $this->conn->query($sql);
                return $resultset;
            }
        } catch (PDOException $e) {
            $this->error = $e->getMessage();
        }
    }

    function  insertUser($user, $name, $pass, $email, $date){
        try {
            $sql = "INSERT INTO ".TABLE_USER. "(nombre, fnac, email, admin, usuario, password) VALUES ('".$name."','".$date."','".$email."',0,'".$user."','".$pass."')";
            if ($this->conn->exec($sql) === false){
                return false;
            }else {
                if($this->insertaLog("INSERT","Nuevo usuario registrado, ¡Bienvenido, ".$user."!", $sql))
                    return true;
            }
        } catch (PDOException $e) {
            $this->error = $e->getMessage();
        }
    }

    function  insertReserve($usuario, $aula, $fecha, $hora, $motivo){
        try {
            if($hora == 'all'){
                $sql = "SELECT * FROM ".TABLE_RESERVE." WHERE aula = '".$aula."' AND fecha = '".$fecha."'";
                $stmt = $this->conn->query($sql);
                if ($stmt->rowCount() != 0) {
                    return false;
                }
                else
                    $sql = "INSERT INTO ".TABLE_RESERVE. "(usuario, aula, fecha, tramo, motivo) VALUES ('".$usuario."','".$aula."','".$fecha."','8:15','".$motivo."'),('".$usuario."','".$aula."','".$fecha."','9:15','".$motivo."'),('".$usuario."','".$aula."','".$fecha."','10:15','".$motivo."'),('".$usuario."','".$aula."','".$fecha."','11:45','".$motivo."'),('".$usuario."','".$aula."','".$fecha."','12:45','".$motivo."'),('".$usuario."','".$aula."','".$fecha."','13:45','".$motivo."')";
            }
            else
                $sql = "INSERT INTO ".TABLE_RESERVE. "(usuario, aula, fecha, tramo, motivo) VALUES ('".$usuario."','".$aula."','".$fecha."','".$hora."','".$motivo."')";
            if($this->insertaLog("INSERT",$_SESSION['user']." reservó un aula.", $_SESSION['user'], $sql)) {
                if ($this->conn->exec($sql) === false) {
                    return false;
                } else {
                    return true;
                }
            }
        } catch (PDOException $e) {
            $this->error = $e->getMessage();
        }
    }

    function getAulas()
    {
        try {
            $sql = "SELECT nombre, descripcion, ubicacion, tic, numordenadores, deshabilitada FROM " . TABLE_AULA;
            $resultset = $this->conn->query($sql);
            return $resultset;
        } catch (PDOException $e) {
            $this->error = $e->getMessage();
        }
    }

    function getReservas()
    {
        try {
            if($_SESSION['admin'])
                $sql = "SELECT nreserva, nombre, r.usuario, aula, fecha, tramo, motivo FROM ".TABLE_RESERVE." r,".TABLE_USER." u where r.usuario = u.usuario";
            else
                $sql = "SELECT  aula, fecha, tramo, motivo FROM ".TABLE_RESERVE." WHERE usuario = '".$_SESSION['user']."'";
            $resultset = $this->conn->query($sql);

            return $resultset;
        } catch (PDOException $e) {
            $this->error = $e->getMessage();
        }
    }

    function getUserPrivileges()
    {
        try {
            $sql = "SELECT admin FROM " . TABLE_USER . " WHERE usuario = '" . $_SESSION['user']."'";
            $resultset = $this->conn->query($sql);
            return $resultset->fetch()['admin'];
        } catch (PDOException $e) {
            $this->error = $e->getMessage();
        }
    }

    function deshabilitarClase($nombre)
    {
        try {
            $sql = "UPDATE aula SET deshabilitada='1' WHERE nombre='".$nombre."'";
            var_dump($sql);
            if($this->insertaLog("UPDATE","Deshabilitó el aula ".$nombre, $sql)){
            $resultset = $this->conn->query($sql);
            return $resultset;
        }
        } catch (PDOException $e) {
            $this->error = $e->getMessage();
        }
    }

    function habilitarClase($nombre)
    {
        try {
            $sql = "UPDATE aula SET deshabilitada='0' WHERE nombre='".$nombre."'";
            var_dump($sql);
            if($this->insertaLog("UPDATE","Habilitó el aula ".$nombre, $sql)) {
                $resultset = $this->conn->query($sql);
                return $resultset;
            }
        } catch (PDOException $e) {
            $this->error = $e->getMessage();
        }
    }

    function insertaLog($action,$fullaction,$sql){
        try {
            if($_SESSION['user'] == "")
                $sql = "INSERT INTO log(usuario, shortaccion, fullacion, sqlcommand, datetime) VALUES ('SYSTEM','".$action."','".$fullaction."',\"".$sql."\",NOW())";
            else
                $sql = "INSERT INTO log(usuario, shortaccion, fullacion, sqlcommand, datetime) VALUES ('".$_SESSION['user']."','".$action."','".$fullaction."',\"".$sql."\",NOW())";

            var_dump($sql);
            if ($this->conn->exec($sql) === false){
                return false;
            }else {
                return true;
            }
        } catch (PDOException $e) {
            $this->error = $e->getMessage();
        }
    }

    function getDBScript($rutaDescarga)
    {
        try {
            $sql = "mysqldump -u ".USER." -p".PASS." ".DATABASE." > ".$rutaDescarga;
            $resultset = $this->conn->query($sql);
            return $resultset;
        } catch (PDOException $e) {
            $this->error = $e->getMessage();
        }
    }
}
